src/core/geometry/qgsgeometryutils.cpp

```diff
@@ -1257,3 +1265,533 @@ bool QgsGeometryUtils::transferFirstZValueToPoint( const QgsPointSequence &point
+                                          seg2Start.distance( QgsPoint( chamferEndX, chamferEndY ) ) );
+
+  return true;
+}
+
+bool QgsGeometryUtils::createFillet( const QgsPoint &seg1Start, const QgsPoint &seg1End,
+                                     const QgsPoint &seg2Start, const QgsPoint &seg2End,
+                                     double radius,
+                                     QgsPoint &filletPoint1,
+                                     QgsPoint &filletMidPoint,
+                                     QgsPoint &filletPoint2,
+                                     double epsilon )
+{
+  if ( radius <= 0 )
+    return false;
+
+  // Create fillet arc using the utility function
+  double filletPointsX[3], filletPointsY[3];
+
+  if ( !QgsGeometryUtilsBase::createFillet(
+         seg1Start.x(), seg1Start.y(), seg1End.x(), seg1End.y(),
+         seg2Start.x(), seg2Start.y(), seg2End.x(), seg2End.y(),
+         radius,
+         filletPointsX, filletPointsY,
+         nullptr, nullptr, nullptr, nullptr,
+         nullptr, nullptr, nullptr, nullptr,
+         epsilon ) )
+  {
+    return false;
+  }
+
+  filletPoint1 = interpolatePointOnSegment( filletPointsX[0], filletPointsY[0], seg1Start, seg1End,
+                 seg1Start.distance( QgsPoint( filletPointsX[0], filletPointsY[0] ) ) );
+  filletMidPoint = createPointWithMatchingDimensions( filletPointsX[1], filletPointsY[1], seg1Start );
+  filletPoint2 = interpolatePointOnSegment( filletPointsX[2], filletPointsY[2], seg2Start, seg2End,
+                 seg2Start.distance( QgsPoint( filletPointsX[2], filletPointsY[2] ) ) );
+
+  // Interpolate Z and M for midpoint
+  if ( seg1Start.is3D() && seg1End.is3D() && seg2Start.is3D() && seg2End.is3D() )
+  {
+    filletMidPoint.setZ( ( filletPoint1.z() + filletPoint2.z() ) / 2.0 );
+  }
+  if ( seg1Start.isMeasure() && seg1End.isMeasure() && seg2Start.isMeasure() && seg2End.isMeasure() )
+  {
+    filletMidPoint.setM( ( filletPoint1.m() + filletPoint2.m() ) / 2.0 );
+  }
+
+  return true;
+}
+
+bool QgsGeometryUtils::createFilletArray( const QgsPoint &seg1Start, const QgsPoint &seg1End,
+    const QgsPoint &seg2Start, const QgsPoint &seg2End,
+    double radius,
+    QgsPoint filletPoints[3],
+    double epsilon )
+{
+  QgsPoint p1, p2, p3;
+  if ( createFillet( seg1Start, seg1End, seg2Start, seg2End, radius, p1, p2, p3, epsilon ) )
+  {
+    filletPoints[0] = p1;
+    filletPoints[1] = p2;
+    filletPoints[2] = p3;
+    return true;
+  }
+  return false;
+}
+
+QgsLineString *QgsGeometryUtils::createChamferGeometry(
+  const QgsPoint &seg1Start, const QgsPoint &seg1End,
+  const QgsPoint &seg2Start, const QgsPoint &seg2End,
+  double distance1, double distance2 )
+{
+  QgsPoint chamferStart, chamferEnd;
+  if ( !createChamfer( seg1Start, seg1End, seg2Start, seg2End, distance1, distance2, chamferStart, chamferEnd ) )
+    return nullptr;
+
+  QgsLineString *completeLine = new QgsLineString();
```

Comment:
Use unique_ptr. Also use constructor for QgsLineString which takes a list of points instead of calling addVertex multiple times